src/Intervals.php

```diff
@@ -100,20 +117,157 @@ public static function haveIntersections(ConstraintInterface $a, ConstraintInter
             return true;
         }
 
+        if ($a instanceof MatchNoneConstraint || $b instanceof MatchNoneConstraint) {
+            return false;
+        }
+
         $intersectionIntervals = self::generateIntervals(new MultiConstraint(array($a, $b), true), true);
 
-        return \count($intersectionIntervals['numeric']) > 0 || \count($intersectionIntervals['branches']) > 0;
+        return \count($intersectionIntervals['numeric']) > 0 || $intersectionIntervals['branches']['exclude'] || \count($intersectionIntervals['branches']['names']) > 0;
+    }
+
+    /**
+     * Attempts to optimize a MultiConstraint
+     *
+     * When merging MultiConstraints together they can get very large, this will
+     * compact it by looking at the real intervals covered by all the constraints
+     * and then creates a new constraint containing only the smallest amount of rules
+     * to match the same intervals.
+     *
+     * @return ConstraintInterface
+     */
+    public static function compactConstraint(ConstraintInterface $constraint)
+    {
+        if (!$constraint instanceof MultiConstraint) {
+            return $constraint;
+        }
+
+        $intervals = self::generateIntervals($constraint);
+        $constraints = array();
+        $hasNumericMatchAll = false;
+
+        if (\count($intervals['numeric']) === 1 && (string) $intervals['numeric'][0]->getStart() === (string) Interval::fromZero() && (string) $intervals['numeric'][0]->getEnd() === (string) Interval::untilPositiveInfinity()) {
+            $constraints[] = $intervals['numeric'][0]->getStart();
+            $hasNumericMatchAll = true;
+        } else {
+            $unEqualConstraints = array();
+            for ($i = 0, $count = \count($intervals['numeric']); $i < $count; $i++) {
+                $interval = $intervals['numeric'][$i];
+
+                // if current interval ends with < N and next interval begins with > N we can swap this out for != N
+                // but this needs to happen as a conjunctive expression together with the start of the current interval
+                // and end of next interval, so [>=M, <N] || [>N, <P] => [>=M, !=N, <P] but M/P can be skipped if
+                // they are zero/+inf
+                if ($interval->getEnd()->getOperator() === '<' && $i+1 < $count) {
```

Comment:
inverse the 2 test `$i+1 < $count` is costless than `$interval->getEnd()->getOperator()`